src/Microsoft.AspNetCore.Razor.LanguageServer.Common/VSCodeForegroundDispatcher.cs

```diff
@@ -40,7 +40,17 @@ private ForegroundTaskScheduler()
 
             protected override void QueueTask(Task task) => _tasks.Add(task);
 
-            protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued) => TryExecuteTask(task);
+            protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued)
+            {
+                // If the task was previously queued it means that we're ensuring it's running on our single threaded scheduler.
+                // Otherwise, we can't enforce that behavior and therefore need it to be re-queued before execution.
+                if (taskWasPreviouslyQueued)
```

Comment:
```C#
Task.Factory.StartNew(() => {
    // Note we're not awaiting the async method
    var foo = DoSomethingAsync();
},
CancellationToken.None,
TaskCreationOptions.None,
_foregroundDispatcher.ForegroundScheduler);

private async Task DoSomethingAsync()
{
    await Task.Delay(1000);
   
    // Transitioning back here asks the parent foreground scheduler if it can run the task inline with `taskWasPreviouslyQueued` false.
}
```